docs: Expand rule options schema docs

matwilko opened the pull request:
#### Prerequisites checklist

- [X] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[X] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[] Other, please explain:

#### What changes did you make? (Give an overview)

While reading the documentation on how rules are configured and specifying options schemas on custom rules, it wasn't clear to me exactly how the two kinds of schema values (object/array) were applied, and I spent a few hours tearing my hair out trying to make a custom object schema work before delving into the code and realising I was supposed to be validating an array!

`RuleValidator.validate` and `getRuleOptionsSchema` were simple enough to understand, but the documentation doesn't really spell out some of the details of how the validation works and the schema(s) are applied. I've tried to expand the documentation substantially in this area to try to make the validation process clearer and highlight some implied consequences and pitfalls. I've also expanded the provided examples to show both valid/invalid potential configurations for the example schema, and added an object schema example, along with an invalid example to highlight the pitfall I fell into.

#### Is there anything you'd like reviewers to focus on?

Not sure if the expanded docs are perhaps too verbose? I was trying to err on the side of spelling things out rather than leave out important details.

<!-- markdownlint-disable-file MD004 -->

matwilko commented:
/easycla

Rec0iL99 commented:
Hi @matwilko, thanks for the suggestion. Please sign the CLA. Thanks.

matwilko commented:
> Please sign the CLA.

No problem, I've already started the process, but I need to go through the corporate contributor path so it might take a little while.

Any feedback in the meantime would be appreciated :)

mdjermanovic commented on docs/src/extend/custom-rules.md:
````diff
@@ -645,6 +668,66 @@ module.exports = {
 };
 ```
 
+And as an example for the full JSON Schema object form:
````
This sounds like the following will be the same schema, just in the JSON Schema object form, but it isn't the same. Perhaps it would be better to really show how the schema for `yoda`  rule options would look like in the JSON Schema object form?

matwilko replied:
Sure, I was trying to show that there's more/different stuff you can do with the object schema, but I can see how it looks confusing.

I can split it into two examples, one explicitly showing the array-equivalent object schema, and a second showing off how the object schema can diverge in behaviour?

mdjermanovic commented:
> Not sure if the expanded docs are perhaps too verbose?

No, I think the clarifications are very helpful, thanks!

matwilko commented on tests/lib/config/flat-config-helpers.js:
```diff
@@ -99,4 +102,125 @@ describe("Config Helpers", () => {
+                meta: { schema: [{ enum: [1] }] }
+            });
+
+            assert.isTrue(ajv.validate(outputSchema, [0]));
+            assert.isFalse(ajv.validate(outputSchema, [1]));
```
I tried to write the tests in such a way that it's testing the overall contract of the method, i.e. "given this rule, return _some_ JSON schema (or null) that will validate its sliced options array in the documented way". We don't actually care about the specific shape of the schema returned, we care about what behaviour that schema exhibits.

Sending the output schema through ajv and validating that it accepts/rejects values that we know will be valid/invalid allows the exact shape of the schema to be modified without changing the test. The test will still be testing "does the method give me a schema that will validate the options in the way we expect", but it allows flexibility of implementation.

The `$ref` PR is an example of this (ignoring whether that change actually happens) - the structure of the JSON Schema is being changed, but the test is still testing the actual behaviour of whatever structure is generated. A smaller example might be eliminating `minItems: 0`, since it's technically redundant - the behaviour of the schema would be the same, but a test asserting that it was set would start failing.

matwilko replied:
Tests removed from PR.

matwilko commented on tests/lib/config/flat-config-helpers.js:
```diff
@@ -99,4 +102,125 @@ describe("Config Helpers", () => {
+                it("should allow an empty array", () => {
+                    const outputSchema = getRuleOptionsSchema(testSchema);
+
+                    assert.isTrue(ajv.validate(outputSchema, []));
```
I think there's a case to be made of testing the higher level option validation including severity, and I'm happy to add some tests up there for that. But I think it's still valuable to test this smaller unit of the validation process in this way - as noted above, the tests as written aren't testing the exact shape of the schema generated, but the contract of "given a rule, this method will give us back _some_ schema that will validate the sliced options array".

If the objection is that ajv might get swapped out for something else in the future, necessitating reworking these tests, I think that means we should instead push the ajv dependency down into this method and change its contract to "given a rule, return me a function that will validate the sliced options array". That isolates the ajv dependency and makes the testing strategy clearer (i.e. we're getting back a black box, so let's test its behaviour since we can't see the structure)

nzakas replied:
Agree with @mdjermanovic here: the interface to test should not be `ajv`.

matwilko replied:
Tests removed from PR.

bmish commented on docs/src/extend/custom-rules.md:
```diff
@@ -614,17 +614,40 @@ You can also access comments through many of `sourceCode`'s methods using the `i
 Rules may export a `schema` property, which is a [JSON Schema](https://json-schema.org/) format description of a rule's options which will be used by ESLint to validate configuration options and prevent invalid or unexpected inputs before they are passed to the rule in `context.options`.
 
-There are two formats for a rule's exported `schema`:
+If `schema` is not specified, then any options (besides severity) supplied to the rule are passed on verbatim without validation.
```
Just want to call out that some of this behavior will be changing in the next major release v9 due to [RFC-85](https://github.com/eslint/rfcs/tree/main/designs/2021-schema-object-rules) ([draft PR](https://github.com/eslint/eslint/pull/16614)).


>- Requiring rules with options to specify schemas (removing support for rules with options that are missing schemas)
>- Requiring rule implementations to use the object-style format (removing support for the deprecated function-style format)

It would ideal if we can take that into account here by:
* Avoid providing recommendations that conflict with the upcoming requirements
* Write the documentation in such a way that it can be easily updated when the new requirements come into effect

matwilko replied:
> Avoid providing recommendations that conflict with the upcoming requirements

I don't think the docs as written now _recommend_ not providing a schema, it just explains the consequence of not providing one. I can update the line to note that it is explicitly _not_ recommended, and that this option will be removed in v9. Once v9 drops, the change to the docs should be as simple as replacing that paragraph to explain that if you don't provide a `schema` object, your rule will not be permitted to have any options specified (equivalent to `[]`)?

> Requiring rule implementations to use the object-style format (removing support for the deprecated function-style format)

I don't think these docs have any conflict with that? They explain `schema`/`meta.schema`, but that change is eliminating the `const rule = function(context, options) { ... }; rule.schema = { ... };` form afaict, which is no longer documented at all?

matwilko replied:
Updated docs to:

>If `schema` is not specified, then any options (besides severity) supplied to the rule are passed on verbatim without validation. This is a legacy behaviour though, and all rules are encouraged to specify a schema to aid developers when configuring your rules. Note that in v9, this behaviour will be deprecated entirely, and not specifying `schema` will force the rule to have no options (equivalent to `[]`) - see the [Require schemas and object-style rules](https://github.com/eslint/rfcs/blob/main/designs/2021-schema-object-rules/README.md) RFC.

Does this cover the bases you wanted to cover?

matwilko replied:
Also noting in the PR you linked:

> Throw with no-op schema schema: {}

Would it be helpful to add some further rule schema validations to catch obvious cases that will _always_ fail, like a top-level `type: 'object/number/string'`? Could also recurse into `anyOf`/`oneOf`/`allOf` cases and do the same check - i.e. the first `type` to be tested always needs to be an `array`?

Not necessarily suggesting adding those validations in this PR, though I'm happy to take a stab at it separately?

bmish replied:
Your documentation tweaks sound reasonable, thanks!

As for catching more bad schemas, I certainly wouldn't mind that, but out of scope for my initial [PR](https://github.com/eslint/eslint/pull/16614) implementing that [RFC](https://github.com/eslint/rfcs/tree/main/designs/2021-schema-object-rules). I think we could consider another breaking change PR with more validation though.

matwilko commented:
Added a note that the supported JSON Schema version is Draft-04 and that later additions like `if` and `$data` aren't available.

Also updated the jsdoc of `getRuleOptionsSchema` to properly reflect types.

matwilko commented:
My understanding is that currently all new doc updates are paused until #17225 is completed, so I'll leave the PR open but effectively paused for now, and rebase after the doc freeze is done.

Rec0iL99 commented:
> My understanding is that currently all new doc updates are paused until #17225 is completed, so I'll leave the PR open but effectively paused for now, and rebase after the doc freeze is done.

Waiting for the doc freeze to end.

ollie-iterators commented:
https://github.com/eslint/eslint/issues/17225 has been closed.

nzakas commented on docs/src/extend/custom-rules.md:
```diff
@@ -614,37 +614,161 @@ You can also access comments through many of `sourceCode`'s methods using the `i
 Rules may export a `schema` property, which is a [JSON Schema](https://json-schema.org/) format description of a rule's options which will be used by ESLint to validate configuration options and prevent invalid or unexpected inputs before they are passed to the rule in `context.options`.
 
-There are two formats for a rule's exported `schema`:
+If `schema` is not specified, then any options (besides severity) supplied to the rule are passed on verbatim without validation. This is a legacy behaviour though, and all rules are encouraged to specify a schema to aid developers when configuring your rules. Note that in v9, this behaviour will be deprecated entirely, and not specifying `schema` will force the rule to have no options (equivalen ...
```
```suggestion
**Note:** Prior to ESLint v9.0.0, rules without a `schema` are passed their options directly from the config without any validation. In ESLint v9.0.0 and later, rules without schemas will receive an empty array as `context.options`. See the [Require schemas and object-style rules RFC](https://github.com/eslint/rfcs/blob/main/designs/2021-schema-object-rules/README.md) for further details.
```

matwilko replied:
I'm not sure your suggestion quite covers the proposed strictness in the [RFC](https://github.com/eslint/rfcs/blob/47b9bec8b5d621a70209b29d4a7c237b43d16e33/designs/2021-schema-object-rules/README.md?plain=1#L32C1-L32C1).

I think the way you've phrased it implies that even if a user specifies options for a rule without a schema, they will be ignored and an empty array passed. But the RFC proposes something stricter, i.e. it is an error for a user to specify any options if a rule doesn't provide a schema.

I know the original version doesn't explicitly spell out this consequence either - should we perhaps highlight the consequence of this breaking change for existing schema-less rules? Perhaps the following:

> **Note:** Prior to ESLint v9.0.0, rules without a `schema` are passed their options directly from the config without any validation. In ESLint v9.0.0 and later, rules without schemas will be treated as if they specified a schema of `[]` - this will cause errors for users that specify options for your rule when they upgrade to ESLint v9.0.0+. See the [Require schemas and object-style rules RFC](https://github.com/eslint/rfcs/blob/main/designs/2021-schema-object-rules/README.md) for further details.

matwilko replied:
Yep, even more short and sweet. Will merge in now - I'm just making these changes directly on the branch rather than committing the suggestions in the GH UI 😄

nzakas replied:
That's fine, just make sure the conversations are marked as "Resolved" so we know what's been done.

bmish replied:
Sorry late reply but it looks good.

nzakas commented on docs/src/extend/custom-rules.md:
```diff
@@ -614,37 +614,161 @@ You can also access comments through many of `sourceCode`'s methods using the `i
-There are two formats for a rule's exported `schema`:
+If `schema` is not specified, then any options (besides severity) supplied to the rule are passed on verbatim without validation. This is a legacy behaviour though, and all rules are encouraged to specify a schema to aid developers when configuring your rules. Note that in v9, this behaviour will be deprecated entirely, and not specifying `schema` will force the rule to have no options (equivalen ...
+
+When validating a rule's options, there are 3 main preparation/validation steps:
```
```suggestion
When validating a rule's options, there are three steps:
```

nzakas commented on docs/src/extend/custom-rules.md:
```diff
@@ -614,37 +614,161 @@ You can also access comments through many of `sourceCode`'s methods using the `i
 
-1. A full JSON Schema object describing all possible options the rule accepts.
-2. An array of JSON Schema objects for each optional positional argument.
+1. If the options are not an array, the value is wrapped into an array (e.g. `"off"` becomes `["off"]`)
+2. ESLint will validate the first element of the options array as a severity (`"off"`, `"warn"`, `"error"`, `0`, `1`, `2`)
+3. If the rule is enabled, then the remaining elements of the array are sliced and the rule's schema validation is run on this sliced options array (e.g. `["warn", "never", { someOption: 5 }]` results in the schema being applied to `["never", { someOption: 5 }]`)
```
You are mixing up terms here. "options" means anything passed in the config except the severity, so we need to clean this up a bit.

```suggestion
1. If the rule config is not an array, then the value is wrapped into an array (e.g. `"off"` becomes `["off"]`); if the rule config is an array then it is used directly.
2. ESLint validates the first element of the rule config array as a severity (`"off"`, `"warn"`, `"error"`, `0`, `1`, `2`).
3. If the rule is enabled, then the other elements of the array (everything except the severity) are copied an options array (into the `context.options` property) and the rule's schema validation is run on this array (e.g. `["warn", "never", { someOption: 5 }]` results in the schema being applied to `["never", { someOption: 5 }]`).
```

nzakas commented on docs/src/extend/custom-rules.md:
```diff
@@ -614,37 +614,161 @@ You can also access comments through many of `sourceCode`'s methods using the `i
+3. If the rule is enabled, then the remaining elements of the array are sliced and the rule's schema validation is run on this sliced options array (e.g. `["warn", "never", { someOption: 5 }]` results in the schema being applied to `["never", { someOption: 5 }]`)
 
-In both cases, these should exclude the [severity](../use/configure/rules#rule-severities), as ESLint automatically validates this first.
+Note that this means that the rule schema cannot see or validate the severity, it only sees the array elements _after_ the severity.
```
```suggestion
Note: this means the rule schema cannot validate the severity., The rule schema only validates the array elements _after_ the severity in a rule config.
```

matwilko replied:
Nice. I'm also going to add an extra sentence on the end to really spell it out:

> There is no way for a rule to know what severity it is configured at.

nzakas commented on docs/src/extend/custom-rules.md:
```diff
@@ -614,37 +614,161 @@ You can also access comments through many of `sourceCode`'s methods using the `i
+There are two formats for a rule's exported `schema`:
+
+* An array of JSON Schema objects
+    * Each element will be checked against the same position in the sliced options array
+    * If the sliced options array has fewer elements than there are schemas, then the unmatched schemas are ignored
+    * If the sliced options array has more elements than there are schemas, the validation fails
```
"sliced options" won't make sense to most folks.

```suggestion
    * Each element will be checked against the same position in the `context.options` array.
    * If the `context.options` array has fewer elements than there are schemas, then the unmatched schemas are ignored.
    * If the `context.options` array has more elements than there are schemas, then the validation fails.
```

nzakas commented on docs/src/extend/custom-rules.md:
```diff
@@ -614,37 +614,161 @@ You can also access comments through many of `sourceCode`'s methods using the `i
+    * Each element will be checked against the same position in the sliced options array
+    * If the sliced options array has fewer elements than there are schemas, then the unmatched schemas are ignored
+    * If the sliced options array has more elements than there are schemas, the validation fails
+    * Two important consequences:
+        * Empty rule options beyond severity are always valid (e.g. `["error"]` will be a valid rule option)
+        * An empty array _enforces no options_ for the rule
```
I've re-read this a few times and I'm still not clear on what these lines are trying to say. 

Consequences of what?

Is the first bullet saying that you are not required to pass options beyond severity?

For the second bullet, what is enforcing no options? An empty array where?

matwilko replied:
Sure, I've tried to clarify the wording on this to highlight the always valid/always error cases a bit more:

>    * There are two important consequences to using this format:
>        * It is _always valid_ for a user to provide no options to your rule (beyond severity)
>        * If you specify an empty array, then it is _always an error_ for a user to provide any options to your rule (beyond severity)

nzakas commented on docs/src/extend/custom-rules.md:
```diff
@@ -614,37 +614,161 @@ You can also access comments through many of `sourceCode`'s methods using the `i
+    * Two important consequences:
+        * Empty rule options beyond severity are always valid (e.g. `["error"]` will be a valid rule option)
+        * An empty array _enforces no options_ for the rule
+* A full JSON Schema object that will validate the sliced options array
```
```suggestion
* A full JSON Schema object that will validate the `context.options` array
```

nzakas commented on docs/src/extend/custom-rules.md:
```diff
@@ -614,37 +614,161 @@ You can also access comments through many of `sourceCode`'s methods using the `i
+        * Empty rule options beyond severity are always valid (e.g. `["error"]` will be a valid rule option)
+        * An empty array _enforces no options_ for the rule
+* A full JSON Schema object that will validate the sliced options array
+    * Be very careful that you are _always_ validating against an array, even if you only expect one options value/object to be supplied
```
```suggestion
    * The schema should assume an array of options to validate even if your rule only accepts one option.
```

nzakas commented on docs/src/extend/custom-rules.md:
````diff
@@ -614,37 +614,161 @@ You can also access comments through many of `sourceCode`'s methods using the `i
+};
+```
+
+Object schemas are more powerful though, and can be much more precise and restrictive in what is permitted. For example, the below schema always requires the first option to be specifed (a number between 0 and 10), but the second option is optional, and can either be an object with some options explicitly set, or `"off"` or `"strict"`.
````
```suggestion
Object schemas can be more precise and restrictive in what is permitted. For example, the below schema always requires the first option to be specified (a number between 0 and 10), but the second option is optional, and can either be an object with some options explicitly set, or `"off"` or `"strict"`.
```

matwilko replied:
I've also updated the example code to note use `yoda` as the rule name, to avoid any confusion with the real rule.

nzakas commented on docs/src/extend/custom-rules.md:
````diff
@@ -614,37 +614,161 @@ You can also access comments through many of `sourceCode`'s methods using the `i
+}
+```
+
+And note the easy pitfall of specifying a schema that will _always_ fail because it doesn't account for the array being validated:
````
```suggestion
Remember, rule options are always an array, so be careful not to specify a schema for a non-array type at the top level. Here's an example schema that will always fail validation:
And note the easy pitfall of specifying a schema that will always fail because it doesn't account for the array being validated:
```

matwilko replied:
I've also added another sentence to really spell out the consequence of doing this wrong:

> If your schema does not specify an array at the top-level, users can _never_ enable your rule, as their configuration will always be invalid when the rule is enabled.

nzakas commented on docs/src/extend/custom-rules.md:
````diff
@@ -614,37 +614,161 @@ You can also access comments through many of `sourceCode`'s methods using the `i
+
+```js
+// Possibly trying to validate ["error", { someOptionalProperty: true }], but:
+// THERE ARE NO POSSIBLE VALID OPTIONS FOR THIS SCHEMA WHEN THE RULE IS ENABLED
````
```suggestion
// When the rule is enabled, this options will always fail validation because the options are an array which doesn't match "object"
```

nzakas reviewed (changes requested):
Thanks for putting this together. I left a bunch of suggestions to clean up the verbiage of the documentation and make it a bit clearer.

I also agree with @mdjermanovic about the code changes, so please take a look at that.

You can optionally remove the code changes and we can deal with the documentation only in this PR and you can resubmit the code changes as a separate PR.

matwilko commented:
Thanks for the feedback @nzakas, much appreciated, and definitely some improvement there 😄

I've removed the added tests, so this is now just a docs update. I'll move the new tests into a separate PR, and I'm also going to refactor the rule schema fetching code a little so that `ajv` is pushed down and the tests will then instead be testing against a function that gets returned (with `ajv` as an implementation detail).

I've highlighted any changes I've made on top of your suggestions inline above.

I also added a note to the JSON Schema Draft-04 bullet to highlight that moving to a later spec is explicitly planned to not happen.

nzakas reviewed (approved):
LGTM. Thanks so much for your work and responding to feedback.
